include/toypp/sharedptr.hpp

```diff
@@ -0,0 +1,224 @@
+#ifndef TOYPP_SHAREDPTR_HPP_
+#define TOYPP_SHAREDPTR_HPP_
+
+#include <atomic>
+#include <functional>
+#include <type_traits>
+#include <utility>
+
+namespace tpp {
+
+namespace details {
+template <typename T>
+struct is_unbounded_array : std::false_type {};
+
+template <typename T>
+struct is_unbounded_array<T[]> : std::true_type {};
+}  // namespace details
+
+template <typename T>
+class WeakPtr;
+
+// TODO support static pointer cast conversions
+template <typename T>
+class SharedPtr {
+  using value_type = std::remove_cv_t<std::remove_reference_t<T>>;
```

Comment:
That should be `using value_type = T;`.  I would be upset if `SharedPtr<const MyType&>` silently behaved as if it were `SharedPtr<MyType>`.  I think there should be
```c++
static_assert(!std::is_reference_v<T>, "SharedPtr does not support reference types");`
```
so that users get a clear error if they specify a reference type.  I don't know if there should also be a `static_assert` for cv-qualifiers because I haven't figured out if `SharedPtr<const volatile MyType>` works as expected.